libraries/standard/http/test/HTTPClient_ReadHeader.c

```diff
@@ -0,0 +1,110 @@
+HTTPStatus_t HTTPClient_ReadHeader( const HTTPResponse_t * pResponse,
+                                    const uint8_t * pHeaderName,
+                                    size_t headerNameLen,
+                                    const uint8_t ** pHeaderValueLoc,
+                                    size_t * pHeaderValueLen )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    HTTPParsingContext_t parsingContext;
+    readHeaderContext_t context =
+    {
+        .pHeaderName     = pHeaderName,
+        .headerNameLen   = headerNameLen,
+        .pHeaderValueLoc = pHeaderValueLoc,
+        .pHeaderValueLen = pHeaderValueLen,
+        .headerFound     = 0u
+    };
+
+    HTTPClient_HeaderParsingCallback_t parsingCallback =
+    {
+        .onHeaderCallback = readHeaderParsingCallback,
+        .pContext         = &context
+    };
+
+    memset( &parsingContext, 0, sizeof( HTTPParsingContext_t ) );
+
+    if( pResponse == NULL )
+    {
+        IotLogError( "Parameter check failed: pResponse is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pResponse->pBuffer == NULL )
+    {
+        IotLogError( "Parameter check failed: pRequestHeaders->pBuffer is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pResponse->bufferLen == 0u )
+    {
+        IotLogError( "Parameter check failed: pRequestHeaders->bufferLen is 0: Buffer len should be > 0." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pHeaderName == NULL )
+    {
+        IotLogError( "Parameter check failed: Input header name is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( headerNameLen == 0u )
+    {
+        IotLogError( "Parameter check failed: Input header name length is 0: pHeaderName should be > 0." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pHeaderValueLoc == NULL )
+    {
+        IotLogError( "Parameter check failed: Output parameter for header value location is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pHeaderValueLen == NULL )
+    {
+        IotLogError( "Parameter check failed: Output parameter for header value length is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else
+    {
+        /* Empty else for MISRA 15.7 compliance. */
+    }
+
+    if( returnStatus == HTTP_SUCCESS )
+    {
+        /* Initialize Parsing context with our callback. */
+        returnStatus = HTTPClient_InitializeParsingContext( &parsingContext,
+                                                            &parsingCallback );
+    }
+
+    if( returnStatus == HTTP_SUCCESS )
+    {
+        returnStatus = HTTPClient_ParseResponse( &parsingContext,
```

Comment:
I did not intend for reading the header to use the same function as parsing a new response. 

By having a separate function we define unique callbacks for http-parser to use just for finding a header. With unique callbacks just for header searching, we can stop the parsing early when a header is found. This increases the efficiency of HTTPClient_ReadHeader().

Reply:
We talked offline for clarification. The separate function is `FindHeaderInResponse` which is optimized for parsing the response buffer only for the single header. 
Have pushed the refactored implementation of `ReadHeader` to call the new function